### DISCUSSION TITLE
QwikJs Framework Support

### DISCUSSION THREAD

**Discussion by elilabes** [chakra-ui/zag#392] [General]
With [Qwik JS](https://qwik.builder.io/) fast gaining support it'd be great to use Zag with it.

I believe that it meets both requirements for prop spreading + strongly typed.

**Comment by anubra266** [comment]
@elilabes I looked into this. And [`preventDefault` handling](https://qwik.builder.io/docs/components/events/#prevent-default) in qwik, is a deal breaker.

**Reply by elilabes** [reply]
Understood! Yeah it is a bit of a different way than usual. 


Out of curiosity, is it possible that the props that zag returns for the different elements where needed simply include `preventdefault:click` in the framework abstraction?

**Comment by kee-oth** [comment]
I'd really like to use Zag.js with Qwik as well. @anubra266, does @elilabes's idea make sense? Can Zag props just return `preventdefault:click` and then that issue is squared away?

**Reply by anubra266** [reply]
No...it's not possible. The prevent default has to be an attribute on the jsx element in qwik. So it can't be part of the various conditional logic that happens in the other event handlers.

**Comment by LazyClicks** [comment]
@anubra266 bit late to this but doesn't this help do it?
https://github.com/BuilderIO/qwik/issues/4496#issuecomment-1682935349

**Comment by gabrielgrant** [comment]
seems there's a proposal to fix this in qwik that is now being actively implemented: https://github.com/BuilderIO/qwik/issues/4496#issuecomment-1775426909

curious if this would be sufficient for qwik support in Zag?

**Comment by gabrielgrant** [comment]
Started looking into this more, but seems it's blocked by https://github.com/BuilderIO/qwik/issues/5540

**Reply by gabrielgrant** [reply]
nvm, seems this is able to work fine: https://github.com/BuilderIO/qwik/issues/5540#issuecomment-1846518950

**Comment by LazyClicks** [comment]
@anubra266 there an upcoming fix for this issue in qwik's upcoming version
[feat(qwik): Experimental support for synchronous QRLs #5508](https://github.com/BuilderIO/qwik/pull/5508)

**Comment by gabrielgrant** [comment]
I've been looking into this a bit more now that there's [an experimental implementation of Qwik's synchronous QRLs](https://github.com/BuilderIO/qwik/pull/5508) (ie event handlers that support preventDefault)

## sync QRL restrictions mean they can't be used directly with Zag

At a glance, the API seems like it should work with Zag: Qwik already supported spreading props and event handlers; the new preventDefault solution just requires wrapping the handlers with `sync()`, which seems like something that could be done in `normalizeProps`

Unfortunately, though, there are very tight restrictions around what can be done in a synchronous handler that make them not directly usable with Zag. Specifically, the functions cannot be imported (ie they must be directly written in-line), and they can't access any component state.

I think the no-import restriction makes them even more limited than [the workaround of putting code in useVisibleTask()](https://github.com/BuilderIO/qwik/issues/4496#issuecomment-1682935349). This workaround was mentioned by @LazyClicks [above](https://github.com/chakra-ui/zag/discussions/392#discussioncomment-6844680), but not sure if anyone really looked into exactly how this could work with Zag? It's not immediately obvious how to make this work with the Zag prop-spreading API, but going to continue investigating

## Zag state machine needs to be serialized

The other (bigger?) issue, is Qwik's need to pause/serialize/resume component state. I'm currently poking around how to best make [Zag's state machine](https://zagjs.com/overview/whats-a-machine) play nice with Qwik's [(de)serialization](https://qwik.builder.io/docs/guides/serialization/) and [resumability](https://qwik.builder.io/docs/concepts/resumable/). 

[Qwik does not currently support serializing custom types](https://github.com/BuilderIO/qwik/issues/2083). In the case of the Zag state machine, though, there dosn't seem to be a lot of internal state in the object:
- current state value
- some "context" (this is only briefly mentioned in the docs, so not yet sure what this is or how it is used)

From a high level, it seems plausible that this state could perhaps be exposed to Qwik's existing scope capture/serialization infrastructure and resumed within the component. I'm still learning about [how Zag's state machine works internally](https://github.com/chakra-ui/zag/blob/main/packages/core/src/machine.ts) to see if that is really possible. 

Of course, even getting Qwik to work with the state machine still has the issue that sync events can't access Qwik-ified state anyway. One possible path forward may be to somehow just persist state machine state to the DOM directly (then the sync handler could access it, and wouldn't need to get Qwik's (de)serialization involved), but haven't fully worked through how that could function exactly.

All in all, still some issues to overcome before Qwik can be natively supported.

## Qwik's support for embedding React components

The other entirely different option is using [Qwik's React integration](https://qwik.builder.io/docs/integrations/react/) to just wrap the React version of Zag/Ark/Park components as islands within a Qwik app. It seems this should work, but is likely not ideal -- [the recommendation is to use "wide islands"](https://qwik.builder.io/docs/integrations/react/#build-wide-islands-not-leaf-nodes), rather than wrapping many small, individual "leaf node" UI components (because each wrapped React component becomes it's own entire React app in the page, with no shared styling etc between)

**Comment by mhevery** [comment]
See my comment https://github.com/BuilderIO/qwik/issues/4496#issuecomment-1848083756 as how it could be achieved.

**Comment by kattsushi** [comment]
Some update on this one ?

**Reply by gabrielgrant** [reply]
AFAICT there is still no Qwik-sanctioned path to Zag support unfortunately

I had started down the path of @mhevery 's suggested approach to refactor the Zag components' event handlers into a sync and async portion. But unfortunately the qwik team subsequently made some design decisions that rendered that approach unworkable.

Specifically: the very narrow API to allow imperative control of `preventDefault()` with `sync()` (implemented by @mhevery here: https://github.com/BuilderIO/qwik/pull/5545 )  is incompatible with Zag, because [it doesn't allow `preventDefault()` to be called from an imported function](https://qwik.dev/docs/cookbook/sync-events/) -- this makes it impossible to extract the component event handling functionality into a common library, which is basically the whole premise of Zag

There's another approach to making `preventDefault()` work that I discussed with @wmertens [here](https://github.com/BuilderIO/qwik/issues/4496#issuecomment-1848236734): eagerly resolve the relevant QRLs on page load.  That could work with basically no changes on the Zag side for an initial implementation. For perf, it would still be better to do the sync-vs-async event handler splitting refactor that @mhevery suggested (though would be curious to benchmark and see how much of a difference it actually makes)

One downside of that approach would be a brief delay before components' event handlers were attached, while the QRLs loaded (similar to the rehydration delay seen with most other frameworks' SSR today, though likely significantly shorter). This could be mitigated if Qwik were to officially support including certain QRLs in the initial bundle. That seemed to me like a more straightforward approach than how `sync()` got implemented. But IIUC @mhevery felt that officially supporting fully-synchronous QRLs without restrictions opened up too much risk/temptation for users to abuse the API and degrade their apps' performance.

It's possible this QRL-preloading approach may still be doable somehow in a kinda hacky way without framework support, but I haven't yet figured out how to eagerly force initial QRL resolution without actually running the handler function (hope to look into this more when I have time, but would be great if someone else can determine if/how this is possible). Even if that is possible, though, I still don't love that would basically have us relying on an internal implementation detail [1] without any indication this is officially supported behavior/some assurance from the team that it wouldn't just change without notice in a future release 

[1]: that the Promise from re-loading an already-loaded QRLs actually resolves synchronously

**Reply by gabrielgrant** [reply]
there's also still a bit of an open question about how the Zag state machine would be serialized, but this seems like it should be doable using @mhevery 's proposal to just put the state into DOM attributes 

The short version off all this is that things haven't really changed since [my status update above](https://github.com/chakra-ui/zag/discussions/392#discussioncomment-7804470)

**Reply by plandem** [reply]
@gabrielgrant looks like its not POC anymore and in beta state:
https://qwik.dev/docs/cookbook/sync-events/

**Reply by LazyClicks** [reply]
> AFAICT there is still no Qwik-sanctioned path to Zag support unfortunately
> 
> I had started down the path of @mhevery 's suggested approach to refactor the Zag components' event handlers into a sync and async portion. But unfortunately the qwik team subsequently made some design decisions that rendered that approach unworkable.
> 
> Specifically: the very narrow API to allow imperative control of `preventDefault()` with `sync()` (implemented by @mhevery here: [QwikDev/qwik#5545](https://github.com/QwikDev/qwik/pull/5545) ) is incompatible with Zag, because [it doesn't allow `preventDefault()` to be called from an imported function](https://qwik.dev/docs/cookbook/sync-events/) -- this makes it impossible to extract the component event handling functionality into a common library, which is basically the whole premise of Zag
> 
> There's another approach to making `preventDefault()` work that I discussed with @wmertens [here](https://github.com/BuilderIO/qwik/issues/4496#issuecomment-1848236734): eagerly resolve the relevant QRLs on page load. That could work with basically no changes on the Zag side for an initial implementation. For perf, it would still be better to do the sync-vs-async event handler splitting refactor that @mhevery suggested (though would be curious to benchmark and see how much of a difference it actually makes)
> 
> One downside of that approach would be a brief delay before components' event handlers were attached, while the QRLs loaded (similar to the rehydration delay seen with most other frameworks' SSR today, though likely significantly shorter). This could be mitigated if Qwik were to officially support including certain QRLs in the initial bundle. That seemed to me like a more straightforward approach than how `sync()` got implemented. But IIUC @mhevery felt that officially supporting fully-synchronous QRLs without restrictions opened up too much risk/temptation for users to abuse the API and degrade their apps' performance.
> 
> It's possible this QRL-preloading approach may still be doable somehow in a kinda hacky way without framework support, but I haven't yet figured out how to eagerly force initial QRL resolution without actually running the handler function (hope to look into this more when I have time, but would be great if someone else can determine if/how this is possible). Even if that is possible, though, I still don't love that would basically have us relying on an internal implementation detail [1] without any indication this is officially supported behavior/some assurance from the team that it wouldn't just change without notice in a future release
> 
> [1]: that the Promise from re-loading an already-loaded QRLs actually resolves synchronously

@mhevery hi misko, is it possible that qwik 2.0 introduces something that helps in this case?

**Reply by wmertens** [reply]
Loading QRLs is easy: call `qrl.resolve()`. So if you have a bunch of handlers that need to be synchronous, you can add a visible task that loads them, and they will be called synchronously on events by qwikloader.

Also, nowadays QRLs have gained `.resolved` which allows synchronous access to the contents once they have been loaded. So you can also manually call qrls synchronously with `qrl.resolved?.(...args)`.

Note that QRLs require the framework to load, so while we could optimize which js chunks to prefetch, they still need to be executed before qrls can be called, and I can't think of a way to do that synchronously.

So, complex sync handlers are already possible, for the price of waking up the container early.
1. given `<div onClick$={() => stuff()} />`
2. replace it with `const handler = $(() => stuff()); useVisibleTask$(() => {handler.resolve()}); return <div onClick$={handler}/>`

You can preload a bunch of handlers in the visible task of course.

**Reply by plandem** [reply]
@anubra266 any directions how to create POC or something? I really would like to see Ark-UI in Qwik, especially since it has no decent UI library right now. Ofc, I could go vanilla-extract or pandacss, but one day when it will require something more complex, it can be a blocker :(

So with these recently ideas that @wmertens posted here, what would be the best starting point for POC? grab vanilla-js example and make it work with qwik?

**Reply by segunadebayo** [reply]
Thanks for the ideas in this thread. I managed to create a PoC for the Qwik integration.

https://github.com/chakra-ui/zag/blob/feat/qwik/examples/qwik-ts/src/routes/index.tsx
https://github.com/chakra-ui/zag/blob/feat/qwik/examples/qwik-ts/src/hooks/use-machine.ts

There are still a lot of things to improve on here. Don't have time to dedicate to completing this

**Reply by anubra266** [reply]
@plandem The controlled context is the only issue in that PoC

**Reply by gabrielgrant** [reply]
quick correction - the second link in @segunadebayo 's comment should be https://github.com/chakra-ui/zag/blob/feat/qwik/examples/qwik-ts/src/hooks/use-machine.ts (not .tsx)

**Comment by anubra266** [comment]
@wmertens @plandem This is what I have for now: https://github.com/chakra-ui/zag/pull/1868

Full qwik context on [`feat/qwik`](https://github.com/chakra-ui/zag/blob/feat/qwik/examples/qwik-ts) branch

Improving the final API is what I'm trying to achieve now. E.g. would be nice to not need to have `initialState` key just for providing initial state

**Reply by osdiab** [reply]
looking forward for this to be ready to use! It would immediately make Qwik much more viable of a framework to build on by having a well-tested and fleshed out UI toolkit (while Qwik UI exists, it's got a long way to go, and Zag/Ark/Panda are the jam).

**Reply by gabrielgrant** [reply]
yes, any updates on the status of this and what would be needed to move it toward being mergeable?

**Reply by anubra266** [reply]
@gabrielgrant mostly this. https://github.com/chakra-ui/zag/pull/1876
Current code may not be reliable. But we basically need to trasform into qrls where neccessary. PRs are welcome. Not much of a qwik expert myself

**Comment by segunadebayo** [comment]
Closing as outdated. Zag now has a vanilla adapter (`@zag-js/vanilla`) that can be used with any framework including Qwik.

**Reply by gabrielgrant** [reply]
while the vanilla implementation is certainly appreciated and can technically be used in Qwik, afaict just implementing this directly seems to go pretty strongly against the qwik philosophy of doing as much work/rendering as possible on the server, and only resuming execution on the client when needed. Am I interpreting correctly this vanilla JS implementation wouldn't really allow for server-side rendering? (probably more a question for @wmertens or @mhevery or other Qwik team members than Chakra folks)

IIUC, to provide the best possible support for qwik's model would require some internal refactoring (splitting handlers into sync portions that need to potentially use `preventDefault` and portions that can be async), but I think even just using @wmertens ' [solution](https://github.com/chakra-ui/zag/discussions/392#discussioncomment-10624219) of eagerly loading QRLs that wrap the whole handlers could allow a decent Qwik implementation (at the expense of some extra code being eagerly loaded)

**Reply by wmertens** [reply]
Actually, it should be ok. Qwik v2 [has `useSerializer$`](https://qwikdev-build-v2.qwik-8nx.pages.dev/docs/core/state/#useserializer--createserializer) which allows transparent custom serialization of arbitrary classes. 

So all that is needed is to make a little wrapper that adds the proper serialize and deserialize logic.

V2 is slated for RC soon (but is already running in production at a couple sites)
